rust/headless-client/src/windows.rs

```diff
@@ -0,0 +1,487 @@
+const SERVICE_TYPE: ServiceType = ServiceType::OWN_PROCESS;
+
+pub(crate) struct Signals {
+    sigint: tokio::signal::windows::CtrlC,
+}
+
+impl Signals {
+    pub(crate) fn new() -> Result<Self> {
+        let sigint = tokio::signal::windows::ctrl_c()?;
+        Ok(Self { sigint })
+    }
+
+    pub(crate) fn poll(&mut self, cx: &mut Context) -> Poll<SignalKind> {
+        if self.sigint.poll_recv(cx).is_ready() {
+            return Poll::Ready(SignalKind::Interrupt);
+        }
+        Poll::Pending
+    }
+}
+
+#[derive(clap::Parser, Default)]
+#[command(author, version, about, long_about = None)]
+struct CliIpcService {
+    #[command(subcommand)]
+    command: CmdIpc,
+}
+
+#[derive(clap::Subcommand)]
+enum CmdIpc {
+    #[command(hide = true)]
+    DebugIpcService,
+    IpcService,
+}
+
+impl Default for CmdIpc {
+    fn default() -> Self {
+        Self::IpcService
+    }
+}
+
+// The return value is useful on Linux
+#[allow(clippy::unnecessary_wraps)]
+pub(crate) fn check_token_permissions(_path: &Path) -> Result<()> {
+    // TODO: Make sure the token is only readable by admin / our service user on Windows
+    Ok(())
+}
+
+pub(crate) fn default_token_path() -> std::path::PathBuf {
+    // TODO: System-wide default token path for Windows
+    PathBuf::from("token.txt")
+}
+
+/// Only called from the GUI Client's build of the IPC service
+///
+/// On Windows, this is wrapped specially so that Windows' service controller
+/// can launch it.
+pub(crate) fn run_only_ipc_service() -> Result<()> {
+    let cli = CliIpcService::parse();
+    match cli.command {
+        CmdIpc::DebugIpcService => run_debug_ipc_service(cli),
+        CmdIpc::IpcService => windows_service::service_dispatcher::start(SERVICE_NAME, ffi_service_run).context("windows_service::service_dispatcher failed. This isn't running in an interactive terminal, right?"),
+    }
+}
+
+fn run_debug_ipc_service(cli: CliIpcService) -> Result<()> {
+    crate::debug_command_setup()?;
+    let rt = tokio::runtime::Runtime::new()?;
+    let mut ipc_service = pin!(ipc_listen(cli));
+    let mut signals = Signals::new()?;
+    rt.block_on(async {
+        std::future::poll_fn(|cx| {
+            match signals.poll(cx) {
+                Poll::Ready(SignalKind::Hangup) => {
+                    return Poll::Ready(Err(anyhow::anyhow!(
+                        "Impossible, we don't catch Hangup on Windows"
+                    )));
+                }
+                Poll::Ready(SignalKind::Interrupt) => {
+                    tracing::info!("Caught Interrupt signal");
+                    return Poll::Ready(Ok(()));
+                }
+                Poll::Pending => {}
+            }
+
+            match ipc_service.as_mut().poll(cx) {
+                Poll::Ready(Ok(())) => {
+                    return Poll::Ready(Err(anyhow::anyhow!(
+                        "Impossible, ipc_listencan't return Ok"
+                    )));
+                }
```

Comment:
If you make it return `Result<Infallible>`, you can do:

```suggestion
                Poll::Ready(Ok(infallible)) => match infallible {},
```

Reply:
Included in #5158